notebooks/NonClassicalSearch.ipynb

```diff
@@ -0,0 +1,1027 @@
+      ],
+      "text/plain": [
+       "<IPython.core.display.Markdown object>"
+      ]
+     },
+     "execution_count": 18,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "%%python\n",
+    "from notebookUtils import *\n",
+    "pseudocode('Simulated Annealing')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The implementation of the above pseudo code can be viewed [here](https://github.com/aimacode/aima-java/blob/AIMA3e/aima-core/src/main/java/aima/core/search/local/SimulatedAnnealingSearch.java) and it can be visualized [here](http://aimacode.github.io/aima-javascript/4-Beyond-Classical-Search/#simulated-annealing). \n",
+    "\n",
+    "Let's find out the probability of accepting a move at 2 different temperatures. As per the algorithm, bad moves are more likely to be accepted at a higher temperature. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 64,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Probability of acceptance of move at lower temperature: 0.36787944117144233\n",
+      "Probability of acceptance of move at higher temperature: 0.951229424500714\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "null"
+      ]
+     },
+     "execution_count": 64,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import aima.core.agent.Action;\n",
+    "import aima.core.search.local.SimulatedAnnealingSearch;\n",
+    "\n",
+    "SimulatedAnnealingSearch<String, Action> search = new SimulatedAnnealingSearch<>(null);\n",
+    "int deltaE = -1; //bad move as ΔE<0\n",
+    "double lowerTemperature = 1.0;\n",
+    "double higherTemperature = 20.0;\n",
+    "\n",
+    "System.out.println(\"Probability of acceptance of move at lower temperature: \" + search.probabilityOfAcceptance(lowerTemperature, deltaE));\n",
+    "System.out.println(\"Probability of acceptance of move at higher temperature: \" + search.probabilityOfAcceptance(higherTemperature, deltaE));"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Local beam search\n",
+    "\n",
+    "This algorithm keeps track of $k$ states rather than just one. It begins with $k$ randomly generated states. At each step, all the successors of all $k$ states are generated. If anyone is a goal then algorithm halts. Otherwise, it selects the $k$ best successors from the complete list and repeats. \n",
+    "\n",
+    "Note that the two algorithms i.e. random-restart search and local beam search are quite different. In a random restart search, each search process runs independently of the others. _In a local beam, useful information is passed among the parallel search threads_. In effect, the algorithm quickly abandons unfruitful searches and moves its resources to where the most progress has been made. "
+   ]
```

Comment:
We should have examples demonstrating the use of simulated annealing and beam search on concrete problems. Take a look at this: http://www.cs.cmu.edu/afs/cs/academic/class/46927-f97/slides/Lec3/sld023.htm